<a href="https://colab.research.google.com/github/pavan03ds/Feature-Engineering/blob/main/1_Handling_categorical_Features_1_One_Hot_Encoding.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#one hot encoding

In [1]:
import pandas as pd
import numpy as np

In [6]:
df = pd.read_csv('/content/drive/MyDrive/Feature Engineering/titanic.csv',usecols=['Sex'])

In [8]:
df.head()

Sex
0    male
1  female
2  female
3  female
4    male

In [9]:
pd.get_dummies(df,drop_first=True)

Sex_male
0           1
1           0
2           0
3           0
4           1
..        ...
886         1
887         0
888         0
889         1
890         1

[891 rows x 1 columns]

In [10]:
df = pd.read_csv('/content/drive/MyDrive/Feature Engineering/titanic.csv',usecols=['Embarked'])

In [11]:
df.head()

Embarked
0        S
1        C
2        S
3        S
4        S

In [12]:
df['Embarked'].unique()

array(['S', 'C', 'Q', nan], dtype=object)

In [13]:
df.dropna(inplace=True)

In [14]:
pd.get_dummies(df,drop_first=True)

Embarked_Q  Embarked_S
0             0           1
1             0           0
2             0           1
3             0           1
4             0           1
..          ...         ...
886           0           1
887           0           1
888           0           1
889           0           0
890           1           0

[889 rows x 2 columns]

#

# One Hot Encoding - variables with many categories

In [17]:
data = pd.read_csv(r"/content/drive/MyDrive/Feature Engineering/mercedesbenz.csv",usecols=['X1','X2','X3','X4','X5','X6'])

In [18]:
data.head()

X1  X2 X3 X4 X5 X6
0  v  at  a  d  u  j
1  t  av  e  d  y  l
2  w   n  c  d  x  j
3  t   n  f  d  x  l
4  v   n  f  d  h  d

In [19]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4209 entries, 0 to 4208
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   X1      4209 non-null   object
 1   X2      4209 non-null   object
 2   X3      4209 non-null   object
 3   X4      4209 non-null   object
 4   X5      4209 non-null   object
 5   X6      4209 non-null   object
dtypes: object(6)
memory usage: 197.4+ KB


In [20]:
#how many categories/lables in each variable has
for col in data.columns:
    print(col, ':' ,len(data[col].unique()),'lables')

X1 : 27 lables
X2 : 44 lables
X3 : 7 lables
X4 : 4 lables
X5 : 29 lables
X6 : 12 lables


In [21]:
#how many varibles will be there after one hot encoding?
pd.get_dummies(data,drop_first=True).shape

(4209, 117)

we can see that from just 6 initial categories variables,we get 117 new variables

Instead...
we would make a binary variable for each of the 10 most frequent variables only.
This is equivalent to grouping all the other lables under new category, that in this case wiil be dropped.Thus, the 10 new dummy variables indicate if one of the 10 most frequent lables is present (1) or not (0) for a particular obsevation.

In [22]:
data.X2.value_counts().sort_values(ascending=False).head(15)

as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
aq      63
ay      54
a       47
t       29
k       25
Name: X2, dtype: int64

In [23]:
#values with top 10 count of occurence of values in X2 
top10 = [x for x in data.X2.value_counts().sort_values(ascending=False).head(10).index]
top10

['as', 'ae', 'ai', 'm', 'ak', 'r', 'n', 's', 'f', 'e']

In [24]:
for label in top10:
    data[label]=np.where(data['X2']==label,1,0)
data[['X2']+top10].head(30)

X2  as  ae  ai  m  ak  r  n  s  f  e
0   at   0   0   0  0   0  0  0  0  0  0
1   av   0   0   0  0   0  0  0  0  0  0
2    n   0   0   0  0   0  0  1  0  0  0
3    n   0   0   0  0   0  0  1  0  0  0
4    n   0   0   0  0   0  0  1  0  0  0
5    e   0   0   0  0   0  0  0  0  0  1
6    e   0   0   0  0   0  0  0  0  0  1
7   as   1   0   0  0   0  0  0  0  0  0
8   as   1   0   0  0   0  0  0  0  0  0
9   aq   0   0   0  0   0  0  0  0  0  0
10   r   0   0   0  0   0  1  0  0  0  0
11   e   0   0   0  0   0  0  0  0  0  1
12  ai   0   0   1  0   0  0  0  0  0  0
13   e   0   0   0  0   0  0  0  0  0  1
14  as   1   0   0  0   0  0  0  0  0  0
15  ak   0   0   0  0   1  0  0  0  0  0
16   e   0   0   0  0   0  0  0  0  0  1
17  as   1   0   0  0   0  0  0  0  0  0
18   m   0   0   0  1   0  0  0  0  0  0
19  ak   0   0   0  0   1  0  0  0  0  0
20   m   0   0   0  1   0  0  0  0  0  0
21  as   1   0   0  0   0  0  0  0  0  0
22   a   0   0   0  0   0  0  0  0  0  0
23   m   0   0   0  1   0  0  0  0  0  0
24  ak   0   0   0  0   1  0  0  0  0  0
25   m   0   0   0  1   0  0  0  0  0  0
26  as   1   0   0  0   0  0  0  0  0  0
27  as   1   0   0  0   0  0  0  0  0  0
28  as   1   0   0  0   0  0  0  0  0  0
29   m   0   0   0  1   0  0  0  0  0  0

In [27]:
#getting whole set of dummy variables for all categorical variables
def one_hot_topx(df ,variable ,top_x_lables):
    
    for label in top_x_lables:
        df[variable+'_'+label] = np.where(df[variable]==label,1,0)
    
#reading data again
data = pd.read_csv(r"/content/drive/MyDrive/Feature Engineering/mercedesbenz.csv",usecols=['X1','X2','X3','X4','X5','X6'])

#encode X2 into top most frequent categories 
one_hot_topx(data,'X2',top10)
data.head()

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  X2_ak  X2_r  X2_n  X2_s  \
0  v  at  a  d  u  j      0      0      0     0      0     0     0     0   
1  t  av  e  d  y  l      0      0      0     0      0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0      0     0     1     0   
3  t   n  f  d  x  l      0      0      0     0      0     0     1     0   
4  v   n  f  d  h  d      0      0      0     0      0     0     1     0   

   X2_f  X2_e  
0     0     0  
1     0     0  
2     0     0  
3     0     0  
4     0     0

In [28]:
#10 most frequent categories of X1
top10 = [x for x in data['X1'].value_counts().sort_values(ascending=False).head(10).index]
top10

['aa', 's', 'b', 'l', 'v', 'r', 'i', 'a', 'c', 'o']

In [29]:
#encode X1 into top most frequent categories 
one_hot_topx(data ,'X1',top10)
data.head(10)

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  ...  X1_aa  X1_s  X1_b  X1_l  \
0  v  at  a  d  u  j      0      0      0     0  ...      0     0     0     0   
1  t  av  e  d  y  l      0      0      0     0  ...      0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0  ...      0     0     0     0   
3  t   n  f  d  x  l      0      0      0     0  ...      0     0     0     0   
4  v   n  f  d  h  d      0      0      0     0  ...      0     0     0     0   
5  b   e  c  d  g  h      0      0      0     0  ...      0     0     1     0   
6  r   e  f  d  f  h      0      0      0     0  ...      0     0     0     0   
7  l  as  f  d  f  j      1      0      0     0  ...      0     0     0     1   
8  s  as  e  d  f  i      1      0      0     0  ...      0     1     0     0   
9  b  aq  c  d  f  a      0      0      0     0  ...      0     0     1     0   

   X1_v  X1_r  X1_i  X1_a  X1_c  X1_o  
0     1     0     0     0     0     0  
1     0     0     0     0     0     0  
2     0     0     0     0     0     0  
3     0     0     0     0     0     0  
4     1     0     0     0     0     0  
5     0     0     0     0     0     0  
6     0     1     0     0     0     0  
7     0     0     0     0     0     0  
8     0     0     0     0     0     0  
9     0     0     0     0     0     0  

[10 rows x 26 columns]

In [30]:
#10 most frequent categories of X3
top10 = [x for x in data['X3'].value_counts().sort_values(ascending=False).head(10).index]
top10

['c', 'f', 'a', 'd', 'g', 'e', 'b']

In [31]:
#encode X1 into top most frequent categories
one_hot_topx(data,'X3',top10)

In [32]:
data.head(10)

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  ...  X1_a  X1_c  X1_o  X3_c  \
0  v  at  a  d  u  j      0      0      0     0  ...     0     0     0     0   
1  t  av  e  d  y  l      0      0      0     0  ...     0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0  ...     0     0     0     1   
3  t   n  f  d  x  l      0      0      0     0  ...     0     0     0     0   
4  v   n  f  d  h  d      0      0      0     0  ...     0     0     0     0   
5  b   e  c  d  g  h      0      0      0     0  ...     0     0     0     1   
6  r   e  f  d  f  h      0      0      0     0  ...     0     0     0     0   
7  l  as  f  d  f  j      1      0      0     0  ...     0     0     0     0   
8  s  as  e  d  f  i      1      0      0     0  ...     0     0     0     0   
9  b  aq  c  d  f  a      0      0      0     0  ...     0     0     0     1   

   X3_f  X3_a  X3_d  X3_g  X3_e  X3_b  
0     0     1     0     0     0     0  
1     0     0     0     0     1     0  
2     0     0     0     0     0     0  
3     1     0     0     0     0     0  
4     1     0     0     0     0     0  
5     0     0     0     0     0     0  
6     1     0     0     0     0     0  
7     1     0     0     0     0     0  
8     0     0     0     0     1     0  
9     0     0     0     0     0     0  

[10 rows x 33 columns]

In [33]:
#10 most frequent categories of X4
top10 = [x for x in data.X4.value_counts().sort_values(ascending=False).head(10).index]
top10

['d', 'a', 'b', 'c']

In [34]:
#encode X4 into top most frequent categories
one_hot_topx(data,'X4',top10)

In [35]:
data.head()

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  ...  X3_f  X3_a  X3_d  X3_g  \
0  v  at  a  d  u  j      0      0      0     0  ...     0     1     0     0   
1  t  av  e  d  y  l      0      0      0     0  ...     0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0  ...     0     0     0     0   
3  t   n  f  d  x  l      0      0      0     0  ...     1     0     0     0   
4  v   n  f  d  h  d      0      0      0     0  ...     1     0     0     0   

   X3_e  X3_b  X4_d  X4_a  X4_b  X4_c  
0     0     0     1     0     0     0  
1     1     0     1     0     0     0  
2     0     0     1     0     0     0  
3     0     0     1     0     0     0  
4     0     0     1     0     0     0  

[5 rows x 37 columns]

In [40]:
#10 most frequent categories of X5
top10 = [x for x in data.X5.value_counts().sort_values().head(10).index]
top10

['u', 'g', 'y', 'h', 'x', 'f', 'o', 'ah', 'aa', 'j']

In [42]:
#encode X5 into top most frequent categories
one_hot_topx(data,'X5',top10)

In [43]:
data.head()

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  ...  X5_u  X5_g  X5_y  X5_h  \
0  v  at  a  d  u  j      0      0      0     0  ...     1     0     0     0   
1  t  av  e  d  y  l      0      0      0     0  ...     0     0     1     0   
2  w   n  c  d  x  j      0      0      0     0  ...     0     0     0     0   
3  t   n  f  d  x  l      0      0      0     0  ...     0     0     0     0   
4  v   n  f  d  h  d      0      0      0     0  ...     0     0     0     1   

   X5_x  X5_f  X5_o  X5_ah  X5_aa  X5_j  
0     0     0     0      0      0     0  
1     0     0     0      0      0     0  
2     1     0     0      0      0     0  
3     1     0     0      0      0     0  
4     0     0     0      0      0     0  

[5 rows x 47 columns]

In [44]:
#10 most frequent categories of X6
top10 = [x for x in data.X6.value_counts().sort_values().head(10).index]
top10

['e', 'f', 'b', 'c', 'k', 'h', 'a', 'l', 'i', 'd']

In [45]:
#encode X5 into top most frequent categories
one_hot_topx(data,'X6',top10)

In [46]:
data.head()

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  ...  X6_e  X6_f  X6_b  X6_c  \
0  v  at  a  d  u  j      0      0      0     0  ...     0     0     0     0   
1  t  av  e  d  y  l      0      0      0     0  ...     0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0  ...     0     0     0     0   
3  t   n  f  d  x  l      0      0      0     0  ...     0     0     0     0   
4  v   n  f  d  h  d      0      0      0     0  ...     0     0     0     0   

   X6_k  X6_h  X6_a  X6_l  X6_i  X6_d  
0     0     0     0     0     0     0  
1     0     0     0     1     0     0  
2     0     0     0     0     0     0  
3     0     0     0     1     0     0  
4     0     0     0     0     0     1  

[5 rows x 57 columns]

In [47]:
data.tail()

X1  X2 X3 X4  X5 X6  X2_as  X2_ae  X2_ai  X2_m  ...  X6_e  X6_f  X6_b  \
4204  s  as  c  d  aa  d      1      0      0     0  ...     0     0     0   
4205  o   t  d  d  aa  h      0      0      0     0  ...     0     0     0   
4206  v   r  a  d  aa  g      0      0      0     0  ...     0     0     0   
4207  r   e  f  d  aa  l      0      0      0     0  ...     0     0     0   
4208  r  ae  c  d  aa  g      0      1      0     0  ...     0     0     0   

      X6_c  X6_k  X6_h  X6_a  X6_l  X6_i  X6_d  
4204     0     0     0     0     0     0     1  
4205     0     0     1     0     0     0     0  
4206     0     0     0     0     0     0     0  
4207     0     0     0     0     1     0     0  
4208     0     0     0     0     0     0     0  

[5 rows x 57 columns]

In [48]:
#Dropping the original columns
data.drop(['X1','X2','X3','X4','X5','X6'],axis=1,inplace=True)

In [49]:
data.head()

X2_as  X2_ae  X2_ai  X2_m  X2_ak  X2_r  X2_n  X2_s  X2_f  X2_e  ...  X6_e  \
0      0      0      0     0      0     0     0     0     0     0  ...     0   
1      0      0      0     0      0     0     0     0     0     0  ...     0   
2      0      0      0     0      0     0     1     0     0     0  ...     0   
3      0      0      0     0      0     0     1     0     0     0  ...     0   
4      0      0      0     0      0     0     1     0     0     0  ...     0   

   X6_f  X6_b  X6_c  X6_k  X6_h  X6_a  X6_l  X6_i  X6_d  
0     0     0     0     0     0     0     0     0     0  
1     0     0     0     0     0     0     1     0     0  
2     0     0     0     0     0     0     0     0     0  
3     0     0     0     0     0     0     1     0     0  
4     0     0     0     0     0     0     0     0     1  

[5 rows x 51 columns]

In [50]:
data.tail()

X2_as  X2_ae  X2_ai  X2_m  X2_ak  X2_r  X2_n  X2_s  X2_f  X2_e  ...  \
4204      1      0      0     0      0     0     0     0     0     0  ...   
4205      0      0      0     0      0     0     0     0     0     0  ...   
4206      0      0      0     0      0     1     0     0     0     0  ...   
4207      0      0      0     0      0     0     0     0     0     1  ...   
4208      0      1      0     0      0     0     0     0     0     0  ...   

      X6_e  X6_f  X6_b  X6_c  X6_k  X6_h  X6_a  X6_l  X6_i  X6_d  
4204     0     0     0     0     0     0     0     0     0     1  
4205     0     0     0     0     0     1     0     0     0     0  
4206     0     0     0     0     0     0     0     0     0     0  
4207     0     0     0     0     0     0     0     1     0     0  
4208     0     0     0     0     0     0     0     0     0     0  

[5 rows x 51 columns]

In [51]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4209 entries, 0 to 4208
Data columns (total 51 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   X2_as   4209 non-null   int64
 1   X2_ae   4209 non-null   int64
 2   X2_ai   4209 non-null   int64
 3   X2_m    4209 non-null   int64
 4   X2_ak   4209 non-null   int64
 5   X2_r    4209 non-null   int64
 6   X2_n    4209 non-null   int64
 7   X2_s    4209 non-null   int64
 8   X2_f    4209 non-null   int64
 9   X2_e    4209 non-null   int64
 10  X1_aa   4209 non-null   int64
 11  X1_s    4209 non-null   int64
 12  X1_b    4209 non-null   int64
 13  X1_l    4209 non-null   int64
 14  X1_v    4209 non-null   int64
 15  X1_r    4209 non-null   int64
 16  X1_i    4209 non-null   int64
 17  X1_a    4209 non-null   int64
 18  X1_c    4209 non-null   int64
 19  X1_o    4209 non-null   int64
 20  X3_c    4209 non-null   int64
 21  X3_f    4209 non-null   int64
 22  X3_a    4209 non-null   int64
 23  X3_d    4209 